src/mlpack/methods/linear_svm/linear_svm_function_impl.hpp

```diff
@@ -0,0 +1,363 @@
+    values(loc) = (*it);
+
+    ++it;
+    ++loc;
+  }
+
+  groundTruth = arma::sp_mat(newLocations, values, groundTruth.n_rows,
+                             groundTruth.n_cols);
+}
+
+template <typename MatType>
+double LinearSVMFunction<MatType>::Evaluate(
+    const arma::mat& parameters)
+{
+  // The objective function is the hinge loss function and it is
+  // calculated over all the training examples.
+
+  // Calculate the loss and regularization terms.
+  // L_i = Σ_i Σ_m max(0, Δ + (w_m x_i + b_m) - (w_{y_i} x_i + b_{y_i}))
+  // where (m != y_i)
+  double loss, regularization, cost;
+
+  // Scores for each class are evaluated.
+  arma::mat scores = parameters * dataset;
+
+  // Make a matrix where each column holds the value of the
+  // correct class Score.
+  arma::mat correctScores = arma::repmat(arma::ones(numClasses).t()
+      * (scores % groundTruth), numClasses, 1);
+
+  // Evaluates the margin and makes (y_i) term 0 so that
+  // they don't have any contribution in the summation.
+  arma::mat margin = scores - correctScores + delta -
+      (delta * groundTruth);
+
+  // The Hinge Loss Function
+  loss = arma::accu(arma::clamp(margin, 0.0, DBL_MAX));
+  loss /= dataset.n_cols;
+
+  // Adding the regularization term.
+  regularization = 0.5 * lambda * arma::dot(parameters,
+      parameters);
+
+  cost = loss + regularization;
+  return cost;
+}
+
+template <typename MatType>
+double LinearSVMFunction<MatType>::Evaluate(
+    const arma::mat& parameters,
+    const size_t firstId,
+    const size_t batchSize)
+{
+  const size_t lastId = firstId + batchSize - 1;
+
+  // Calculate the loss and regularization terms.
+  double loss, regularization, cost;
+
+  arma::mat scores = parameters * dataset;
+  arma::mat correctScores = arma::repmat(arma::ones(numClasses).t()
+      * (scores % groundTruth), numClasses, 1);
+  arma::mat margin = scores - correctScores + delta
+      - (delta * groundTruth);
+
+  // The Hinge Loss Function
+  loss = arma::accu(arma::clamp(margin.cols(firstId, lastId),
+      0.0, DBL_MAX));
+  loss /= batchSize;
+
+  // Adding the regularization term.
+  regularization = 0.5 * lambda * arma::dot(parameters,
+      parameters);
+
+  cost = loss + regularization;
+  return cost;
+}
+
+template <typename MatType>
+template <typename GradType>
+void LinearSVMFunction<MatType>::Gradient(
+    const arma::mat& parameters,
+    GradType& gradient)
+{
+  arma::mat scores = parameters * dataset;
+  arma::mat correctScores = arma::repmat(arma::ones(numClasses).t()
+      * (scores % groundTruth), numClasses, 1);
+  arma::mat margin = scores - correctScores + delta
+      - (delta * groundTruth);
+
+  // For each sample, find the total number of classes where
+  // ( margin > 0 ).
+  arma::mat mask = margin.for_each([](arma::mat::elem_type& val)
+      { val = (val > 0) ? 1: 0; });
+
+  arma::mat difference = groundTruth
+      % (-arma::repmat(arma::sum(mask), numClasses, 1)) + mask;
```

Comment:
The gradient is evaluated as follows:
- Add `x_i` to `w_j` if `margin_i_j`is positive.
- Subtract `x_i` from `w_y_i` for each positive `margin_i_j`.
- Take the average over the size of dataset.
- Add the regularization parameter.

Also, I've updated the comments to address this.

Reply:
Thanks for the clarification.  Did you double-check that my derivation is right?  It might not be. :)  Also, with the new implementation, were you able to reduce the tolerance of the gradient norm check successfully?

Reply:
I made a new test to evaluate the Gradient. So I was able to reduce tolerance to `1e-5` on the new test case, but the old test case was failing sometimes for smaller values of tolerance.